File: app/lib/features/receptionist/domain/receptionist_repository.dart
import 'package:app/core/domain/ids.dart';
import 'package:app/core/domain/phone_number.dart';
import 'package:app/core/result/result.dart';
import 'package:app/features/receptionist/domain/receptionist.dart';

/// One repository interface for the Receptionist Aggregate Root, per
/// docs/15_Technical_Architecture.md §15.15's aggregate map. Every method
/// is Organization- and Branch-scoped, required leading parameters, never
/// optional filters — the same compile-time isolation guarantee as every
/// other Branch-scoped repository (docs/08_System_Architecture.md §8.4).
abstract class ReceptionistRepository {
  /// Invites a new Receptionist for [phoneNumber], assigning them to
  /// exactly one [branchId] (FR-7.1). This write is server-mediated — see
  /// `FirestoreReceptionistRepository`'s doc comment for why the client
  /// cannot perform it directly.
  Future<Result<Receptionist>> invite({
    required OrganizationId organizationId,
    required BranchId branchId,
    required String name,
    required PhoneNumber phoneNumber,
    required String invitedBy,
  });

  /// Lists every Receptionist assigned to [branchId].
  Future<Result<List<Receptionist>>> listByBranch({
    required OrganizationId organizationId,
    required BranchId branchId,
  });

  /// Moves an existing Receptionist, identified by [phoneNumber] (there is
  /// no roster lookup for this to resolve a userId against client-side —
  /// see `FirestoreReceptionistRepository`'s doc comment), to [toBranchId]
  /// within [organizationId] — the revoke-old/create-new model (BR-23).
  /// This write is server-mediated for the same reason [invite] is.
  Future<Result<Receptionist>> reassign({
    required OrganizationId organizationId,
    required String name,
    required PhoneNumber phoneNumber,
    required BranchId toBranchId,
    required String reassignedBy,
  });
}

File: backend/src/features/receptionist/domain/receptionist-repository.ts
import { Receptionist } from './receptionist';

export interface ReassignReceptionistWrite {
  organizationId: string;
  userId: string;
  fromBranchId: string;
  toBranchId: string;
  name: string;
  phoneNumber: string;
  reassignedBy: string;
}

/**
 * A repository-shaped port for the Receptionist Aggregate, per
 * docs/15_Technical_Architecture.md §15.15. Mirrors
 * app/lib/features/receptionist/domain/receptionist_repository.dart's role
 * on the Flutter side.
 */
export interface ReceptionistRepository {
  create(receptionist: Receptionist): Promise<void>;

  findByUserId(organizationId: string, branchId: string, userId: string): Promise<Receptionist | null>;

  /**
   * Atomically revokes the grant at `fromBranchId` and creates a new
   * active grant at `toBranchId` for the same person — the revoke-old/
   * create-new reassignment model (docs/06_Database_Design.md §6.3,
   * BR-23), executed as a single Firestore batch so the two writes
   * commit together or not at all.
   */
  reassign(input: ReassignReceptionistWrite): Promise<Receptionist>;
}
